--- lib/mwan3/mwan3.sh
+++ lib/mwan3/mwan3.sh
@@ -285,6 +285,9 @@
-		if [ -n "${current##*-N mwan3_rules*}" ]; then
+		if [ -n "${current##*-N mwan3_rules$'\n'*}" ]; then
			mwan3_push_update -N mwan3_rules
		fi
+		if [ -n "${current##*-N mwan3_rules_output$'\n'*}" ]; then
+			mwan3_push_update -N mwan3_rules_output
+		fi
 
 		if [ -n "${current##*-N mwan3_hook*}" ]; then
 			mwan3_push_update -N mwan3_hook
@@ -333,6 +336,57 @@
 
 			for chain in custom connected dynamic; do
 				mwan3_push_update -A mwan3_hook \
+					-m mark ! --mark $MMX_DEFAULT/$MMX_MASK \
+					-j mwan3_${chain}_${family}
+			done
+		fi
+
+		if [ -n "${current##*-N mwan3_output_hook*}" ]; then
+			mwan3_push_update -N mwan3_output_hook
+			# do not mangle ipv6 ra service
+			if [ "$IPT" = "$IPT6" ]; then
+				mwan3_push_update -A mwan3_output_hook \
+						  -p ipv6-icmp \
+						  -m icmp6 --icmpv6-type 133 \
+						  -j RETURN
+				mwan3_push_update -A mwan3_output_hook \
+						  -p ipv6-icmp \
+						  -m icmp6 --icmpv6-type 134 \
+						  -j RETURN
+				mwan3_push_update -A mwan3_output_hook \
+						  -p ipv6-icmp \
+						  -m icmp6 --icmpv6-type 135 \
+						  -j RETURN
+				mwan3_push_update -A mwan3_output_hook \
+						  -p ipv6-icmp \
+						  -m icmp6 --icmpv6-type 136 \
+						  -j RETURN
+				mwan3_push_update -A mwan3_output_hook \
+						  -p ipv6-icmp \
+						  -m icmp6 --icmpv6-type 137 \
+						  -j RETURN
+			fi
+			mwan3_push_update -A mwan3_output_hook \
+					  -m mark --mark 0x0/$MMX_MASK \
+					  -j CONNMARK --restore-mark --nfmask "$MMX_MASK" --ctmask "$MMX_MASK"
+			mwan3_push_update -A mwan3_output_hook \
+					  -m mark --mark 0x0/$MMX_MASK \
+					  -j mwan3_ifaces_in
+
+			for chain in custom connected dynamic; do
+				mwan3_push_update -A mwan3_output_hook \
+					-m mark --mark 0x0/$MMX_MASK \
+					-j mwan3_${chain}_${family}
+			done
+
+			mwan3_push_update -A mwan3_output_hook \
+					  -m mark --mark 0x0/$MMX_MASK \
+					  -j mwan3_rules_output
+			mwan3_push_update -A mwan3_output_hook \
+					  -j CONNMARK --save-mark --nfmask "$MMX_MASK" --ctmask "$MMX_MASK"
+
+			for chain in custom connected dynamic; do
+				mwan3_push_update -A mwan3_output_hook \
 					-m mark ! --mark $MMX_DEFAULT/$MMX_MASK \
 					-j mwan3_${chain}_${family}
 			done
@@ -341,9 +395,12 @@
 		if [ -n "${current##*-A PREROUTING -j mwan3_hook*}" ]; then
 			mwan3_push_update -A PREROUTING -j mwan3_hook
 		fi
-		if [ -n "${current##*-A OUTPUT -j mwan3_hook*}" ]; then
-			mwan3_push_update -A OUTPUT -j mwan3_hook
+		if [ -z "${current##*-A OUTPUT -j mwan3_hook*}" ]; then
+			mwan3_push_update -D OUTPUT -j mwan3_hook
 		fi
+		if [ -n "${current##*-A OUTPUT -j mwan3_output_hook*}" ]; then
+			mwan3_push_update -A OUTPUT -j mwan3_output_hook
+		fi
 		mwan3_push_update COMMIT
 		mwan3_push_update ""
 
@@ -784,7 +841,8 @@
 
 mwan3_set_user_iptables_rule()
 {
-	local ipset ipset_src family proto policy src_ip src_port src_iface src_dev
+	local ipset ipset_src ipset_src_local output_ipset_src output_policy output_src_dev
+	local family proto policy src_ip src_port src_iface src_dev
 	local sticky dest_ip dest_port use_policy timeout policy
 	local global_logging rule_logging loglevel rule_policy rule ipv
 
@@ -795,6 +853,7 @@
 	config_get timeout "$1" timeout 600
 	config_get ipset "$1" ipset
 	config_get ipset_src "$1" ipset_src
+	config_get_bool ipset_src_local "$1" ipset_src_local 0
 	config_get proto "$1" proto all
 	config_get src_ip "$1" src_ip
 	config_get src_iface "$1" src_iface
@@ -830,7 +889,10 @@
 	[ -z "$dest_ip" ] && unset dest_ip
 	[ -z "$src_ip" ] && unset src_ip
 	[ -z "$ipset" ] && unset ipset
-	[ -z "$ipset_src" ] && unset ipset_src
+	if [ -z "$ipset_src" ]; then
+		unset ipset_src
+		ipset_src_local=0
+	fi
 	[ -z "$src_port" ] && unset src_port
 	[ -z "$dest_port" ] && unset dest_port
 	if [ "$proto" != 'tcp' ] && [ "$proto" != 'udp' ]; then
@@ -855,6 +917,12 @@
 
 	if [ -n "$ipset_src" ]; then
 		ipset_src="-m set --match-set $ipset_src src"
+	fi
+	output_ipset_src="$ipset_src"
+	output_src_dev="$src_dev"
+	if [ "$ipset_src_local" -eq 1 ]; then
+		unset output_ipset_src
+		unset output_src_dev
 	fi
 
 	if [ -z "$use_policy" ]; then
@@ -900,6 +968,36 @@
 		policy="mwan3_rule_$1"
 	fi
+	output_policy="$policy"
+	if [ "$ipset_src_local" -eq 1 ] && [ $rule_policy -eq 1 ] && [ "$sticky" -eq 1 ]; then
+		output_policy="mwan3_policy_$use_policy"
+	fi
+
 	if [ "$global_logging" = "1" ] && [ "$rule_logging" = "1" ]; then
+		mwan3_push_update -A mwan3_rules_output \
+				  -p "$proto" \
+				  ${src_ip:+-s} $src_ip \
+				  ${output_src_dev:+-i} $output_src_dev \
+				  ${dest_ip:+-d} $dest_ip \
+				  $ipset $output_ipset_src \
+				  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
+				  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
+				  -m mark --mark 0/$MMX_MASK \
+				  -m comment --comment "$1" \
+				  -j LOG --log-level "$loglevel" --log-prefix "MWAN3($1)"
+	fi
+
+	mwan3_push_update -A mwan3_rules_output \
+			  -p "$proto" \
+			  ${src_ip:+-s} $src_ip \
+			  ${output_src_dev:+-i} $output_src_dev \
+			  ${dest_ip:+-d} $dest_ip \
+			  $ipset $output_ipset_src \
+			  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
+			  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
+			  -m mark --mark 0/$MMX_MASK \
+			  -j $output_policy
+
+	if [ "$global_logging" = "1" ] && [ "$rule_logging" = "1" ]; then
 		mwan3_push_update -A mwan3_rules \
 				  -p "$proto" \
 				  ${src_ip:+-s} $src_ip \
@@ -978,11 +1071,15 @@
 		current="$($IPT -S)"$'\n'
 
 
-		if [ -n "${current##*-N mwan3_rules*}" ]; then
+		if [ -n "${current##*-N mwan3_rules$'\n'*}" ]; then
 			mwan3_push_update -N "mwan3_rules"
 		fi
+		if [ -n "${current##*-N mwan3_rules_output$'\n'*}" ]; then
+			mwan3_push_update -N "mwan3_rules_output"
+		fi
 
 		mwan3_push_update -F mwan3_rules
+		mwan3_push_update -F mwan3_rules_output
 
 		config_foreach mwan3_set_user_iptables_rule rule "$ipv"
 
@@ -1027,6 +1124,12 @@
 {
 	mwan3_interface_hotplug_shutdown $1
 	mwan3_track_clean $1
+
+	local IPT
+	for IPT in "$IPT4" "$IPT6"; do
+		[ "$IPT" = "$IPT6" ] && [ $NO_IPV6 -ne 0 ] && continue
+		$IPT -D OUTPUT -j mwan3_output_hook >/dev/null 2>&1 || true
+	done
 }
 
 mwan3_ifup()
