--- lib/mwan3/mwan3.sh
+++ lib/mwan3/mwan3.sh
@@ -930,7 +930,7 @@
 
 mwan3_set_user_iptables_rule()
 {
-	local ipset ipset_src family proto policy src_ip src_port src_iface src_dev
+	local ipset ipset_src ipset_src_local family proto policy src_ip src_port src_iface src_dev
 	local sticky dest_ip dest_port use_policy timeout policy
 	local global_logging rule_logging loglevel rule_policy  rule ipv
 
@@ -941,6 +941,7 @@
 	config_get timeout "$1" timeout 600
 	config_get ipset "$1" ipset
 	config_get ipset_src "$1" ipset_src
+	config_get_bool ipset_src_local "$1" ipset_src_local 0
 	config_get proto "$1" proto all
 	config_get src_ip "$1" src_ip
 	config_get src_iface "$1" src_iface
@@ -964,7 +965,10 @@
 	[ -z "$dest_ip" ] && unset dest_ip
 	[ -z "$src_ip" ] && unset src_ip
 	[ -z "$ipset" ] && unset ipset
-	[ -z "$ipset_src" ] && unset ipset_src
+	if [ -z "$ipset_src" ]; then
+		unset ipset_src
+		ipset_src_local=0
+	fi
 	[ -z "$src_port" ]  && unset src_port
 	[ -z "$dest_port" ]  && unset dest_port
 	if [ "$proto"  != 'tcp' ]  && [ "$proto" != 'udp' ]; then
@@ -1046,6 +1050,32 @@
 				  -m mark ! --mark 0xfc00/0xfc00 \
 				  -j SET --add-set "mwan3_sticky_$rule" src,src
 		policy="mwan3_rule_$1"
+	fi
+	if [ "$ipset_src_local" -eq 1 ]; then
+		if [ "$global_logging" = "1" ] && [ "$rule_logging" = "1" ]; then
+			mwan3_push_update -A mwan3_rules \
+					  ! -i + \
+					  -p "$proto" \
+					  ${src_ip:+-s} $src_ip \
+					  ${dest_ip:+-d} $dest_ip \
+					  $ipset \
+					  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
+					  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
+					  -m mark --mark 0/$MMX_MASK \
+					  -m comment --comment "$1" \
+					  -j LOG --log-level "$loglevel" --log-prefix "MWAN3($1)"
+		fi
+
+		mwan3_push_update -A mwan3_rules \
+				  ! -i + \
+				  -p "$proto" \
+				  ${src_ip:+-s} $src_ip \
+				  ${dest_ip:+-d} $dest_ip \
+				  $ipset \
+				  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
+				  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
+				  -m mark --mark 0/$MMX_MASK \
+				  -j $policy
 	fi
 	if [ "$global_logging" = "1" ] && [ "$rule_logging" = "1" ]; then
 		mwan3_push_update -A mwan3_rules \
