--- lib/mwan3/mwan3.sh
+++ lib/mwan3/mwan3.sh
@@ -930,7 +930,7 @@
 
 mwan3_set_user_iptables_rule()
 {
-	local ipset family proto policy src_ip src_port src_iface src_dev
+	local ipset ipset_src family proto policy src_ip src_port src_iface src_dev
 	local sticky dest_ip dest_port use_policy timeout policy
 	local global_logging rule_logging loglevel rule_policy  rule ipv
 
@@ -940,6 +940,7 @@
 	config_get sticky "$1" sticky 0
 	config_get timeout "$1" timeout 600
 	config_get ipset "$1" ipset
+	config_get ipset_src "$1" ipset_src
 	config_get proto "$1" proto all
 	config_get src_ip "$1" src_ip
 	config_get src_iface "$1" src_iface
@@ -963,6 +964,7 @@
 	[ -z "$dest_ip" ] && unset dest_ip
 	[ -z "$src_ip" ] && unset src_ip
 	[ -z "$ipset" ] && unset ipset
+	[ -z "$ipset_src" ] && unset ipset_src
 	[ -z "$src_port" ]  && unset src_port
 	[ -z "$dest_port" ]  && unset dest_port
 	if [ "$proto"  != 'tcp' ]  && [ "$proto" != 'udp' ]; then
@@ -983,6 +985,10 @@
 
 	if [ -n "$ipset" ]; then
 		ipset="-m set --match-set $ipset dst"
+	fi
+
+	if [ -n "$ipset_src" ]; then
+		ipset_src="-m set --match-set $ipset_src src"
 	fi
 
 	if [ -z "$use_policy" ]; then
@@ -1047,7 +1053,7 @@
 				  ${src_ip:+-s} $src_ip \
 				  ${src_dev:+-i} $src_dev \
 				  ${dest_ip:+-d} $dest_ip \
-				  $ipset \
+				  $ipset $ipset_src \
 				  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
 				  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
 				  -m mark --mark 0/$MMX_MASK \
@@ -1060,7 +1066,7 @@
 			  ${src_ip:+-s} $src_ip \
 			  ${src_dev:+-i} $src_dev \
 			  ${dest_ip:+-d} $dest_ip \
-			  $ipset \
+			  $ipset $ipset_src \
 			  ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
 			  ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
 			  -m mark --mark 0/$MMX_MASK \
