--- lib/mwan3/mwan3.sh
+++ lib/mwan3/mwan3.sh
@@ -881,7 +881,7 @@
 
 mwan3_set_user_iptables_rule()
 {
-	local ipset family proto policy src_ip src_port src_iface src_dev
+	local ipset ipset_src family proto policy src_ip src_port src_iface src_dev
 	local sticky dest_ip dest_port use_policy timeout rule policy IPT
 	local global_logging rule_logging loglevel
 
@@ -890,6 +890,7 @@
 	config_get sticky $1 sticky 0
 	config_get timeout $1 timeout 600
 	config_get ipset $1 ipset
+	config_get ipset_src $1 ipset_src
 	config_get proto $1 proto all
 	config_get src_ip $1 src_ip
 	config_get src_iface $1 src_iface
@@ -903,6 +904,7 @@
 	[ -z "$dest_ip" ] && unset dest_ip
 	[ -z "$src_ip" ] && unset src_ip
 	[ -z "$ipset" ] && unset ipset
+	[ -z "$ipset_src" ] && unset ipset_src
 	[ -z "$src_port" ]  && unset src_port
 	[ -z "$dest_port" ]  && unset dest_port
 	[ "$proto"  != 'tcp' ]  && [ "$proto" != 'udp' ] && {
@@ -928,6 +930,10 @@
 		ipset="-m set --match-set $ipset dst"
 	fi
 
+	if [ -n "$ipset_src" ]; then
+		ipset_src="-m set --match-set $ipset_src src"
+	fi
+
 	if [ -n "$use_policy" ]; then
 		if [ "$use_policy" == "default" ]; then
 			policy="MARK --set-xmark $MMX_DEFAULT/$MMX_MASK"
@@ -997,7 +1003,7 @@
 				     ${src_ip:+-s} $src_ip \
 				     ${src_dev:+-i} $src_dev \
 				     ${dest_ip:+-d} $dest_ip\
-				     $ipset \
+				     $ipset $ipset_src \
 				     ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
 				     ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
 				     -m mark --mark 0/$MMX_MASK \
@@ -1010,7 +1016,7 @@
 			     ${src_ip:+-s} $src_ip \
 			     ${src_dev:+-i} $src_dev \
 			     ${dest_ip:+-d} $dest_ip\
-			     $ipset \
+			     $ipset $ipset_src \
 			     ${src_port:+-m} ${src_port:+multiport} ${src_port:+--sports} $src_port \
 			     ${dest_port:+-m} ${dest_port:+multiport} ${dest_port:+--dports} $dest_port \
 			     -m mark --mark 0/$MMX_MASK \
